neuraxle/higher_order_steps.py

```diff
@@ -0,0 +1,117 @@
+    """
+    Base class to represent an higher order step.
+    A higher-order step is just a function that takes an existing step and returns another step that wraps it.
+
+    An higher order step:
+        - cannot be saved
+        - cannot modify the data container
+        - cannot push himself in the execution context
+
+    .. seealso::
+        :class:`~neuraxle.base.MetaStep`
+    """
+    def __init__(self, step: BaseTransformer):
+        MetaStep.__init__(self, wrapped=step)
+        ForceHandleMixin.__init__(self)
+
+    def save(self, context: ExecutionContext, full_dump=False):
+        """
+        Don't save the higher order step. Save the wrapped step only.
+
+        :param context: context to save.
+        :param full_dump: save full dump or not
+        :return: saved wrapped step
+        """
+        if len(context) > 0 and isinstance(context.parents[-1], BaseHigherOrderStep):
+            context.pop()
+        return self.wrapped.save(context, full_dump=full_dump)
+
+    def _will_process(self, data_container: DataContainer, context: ExecutionContext) -> (DataContainer, ExecutionContext):
+        return data_container, context
+
+    def _will_fit(self, data_container: DataContainer, context: ExecutionContext) -> (DataContainer, ExecutionContext):
+        return data_container, context
+
+    def _will_transform_data_container(self, data_container: DataContainer, context: ExecutionContext) -> (DataContainer, ExecutionContext):
+        return data_container, context
+
+    def _will_fit_transform(self, data_container: DataContainer, context: ExecutionContext):
+        return data_container, context
+
+    def _did_fit_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        return data_container
+
+    def _did_process(self, data_container: DataContainer, context: ExecutionContext):
+        return data_container
+
+    def _did_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        return data_container
+
+
+def WithContext(step: BaseTransformer, context: ExecutionContext):
+    """
+    An higher order step to inject a context inside a step.
+    A step with a context forces the pipeline to use that context through handler methods.
+    This is useful for dependency injection because you can register services inside the :class:`ExecutionContext`.
+    It also ensures that the context has registered all the necessary services.
+
+    .. code-block:: python
+
+        context = ExecutionContext(root=tmpdir)
+        context.set_service_locator(ServiceLocator().services) # where services is of type Dict[Type, object]
+
+        p = WithContext(Pipeline([
+            SomeStep().with_assertion_has_services(BaseService)
+        ]), context)
+
+
+    .. seealso::
+        :class:`BaseStep`,
+        :class:`ExecutionContext`,
+        :class:`BaseTransformer`
+    """
+    class StepWithContext(BaseHigherOrderStep):
```

Comment:
```suggestion
    class WithContext(BaseHigherOrderStep):
```
(also note that using a saver will require to have the context as an attribute of WithContext such as `WithContext.context` because the saver will really need to flush it, and when restoring it would use the loading context provided at load time).